agent/agent.go

```diff
@@ -429,24 +416,106 @@ func (a *Agent) processNextTask(ctx context.Context, nt *apimodels.NextTaskRespo
+			oomTracker:                jasper.NewOOMTracker(),
+			unsetFunctionVarsDisabled: nt.UnsetFunctionVarsDisabled,
+		}
+	} else {
+		tc = tcInput
+	}
+	a.jasper.Clear(setupCtx)
+	tc.jasper = a.jasper
+
+	// If the heartbeat aborts the task immediately, we should report that
+	// the task failed during initial task setup.
+	factory, ok := command.GetCommandFactory("setup.initial")
+	if !ok {
+		return tc, setupFailed, shouldExit, errors.New("setup.initial command is not registered")
+	}
+	tc.setCurrentCommand(factory())
+	a.comm.UpdateLastMessageTime()
+
+	var taskConfig *internal.TaskConfig
+	taskConfig, err = a.makeTaskConfig(setupCtx, tc)
+	if err != nil {
+		setupFailed = true
+		err = errors.Wrap(err, "making task config")
+		grip.Error(err)
+		grip.Infof("Task complete: '%s'.", tc.task.ID)
+		tc.logger = client.NewSingleChannelLogHarness("agent.error", a.defaultLogger)
+		shouldExit, err := a.handleTaskResponse(setupCtx, tc, evergreen.TaskSystemFailed, err.Error())
+		return tc, setupFailed, shouldExit, err
+	}
+	tc.setTaskConfig(taskConfig)
+	if err = a.startLogging(agentCtx, tc); err != nil {
+		setupFailed = true
+		err = errors.Wrap(err, "setting up logger producer")
+		grip.Error(err)
+		grip.Infof("Task complete: '%s'.", tc.task.ID)
+		tc.logger = client.NewSingleChannelLogHarness("agent.error", a.defaultLogger)
+		shouldExit, err := a.handleTaskResponse(setupCtx, tc, evergreen.TaskSystemFailed, err.Error())
+		return tc, setupFailed, shouldExit, err
+	}
+
+	if !tc.ranSetupGroup {
+		tc.taskDirectory, err = a.createTaskDirectory(tc)
+		if err != nil {
+			setupFailed = true
+			err = errors.Wrap(err, "creating task directory")
+			grip.Error(err)
+			grip.Infof("Task complete: '%s'.", tc.task.ID)
+			tc.logger.Execution().Error(errors.Wrap(err, "creating task directory"))
+			shouldExit, err := a.handleTaskResponse(setupCtx, tc, evergreen.TaskSystemFailed, err.Error())
+			return tc, setupFailed, shouldExit, err
+		}
 	}
+	tc.taskConfig.WorkDir = tc.taskDirectory
+	tc.taskConfig.Expansions.Put("workdir", tc.taskConfig.WorkDir)
+
+	tc.logger.Task().Infof("Task logger initialized (agent version '%s' from Evergreen build revision '%s').", evergreen.AgentVersion, evergreen.BuildRevision)
+	tc.logger.Execution().Info("Execution logger initialized.")
+	tc.logger.System().Info("System logger initialized.")
+
+	if err := setupCtx.Err(); err != nil {
+		setupFailed = true
+		grip.Error(err)
+		tc.logger.Execution().Infof("Stopping task execution: %s", err)
+		shouldExit, err := a.handleTaskResponse(setupCtx, tc, evergreen.TaskSystemFailed, err.Error())
+		return tc, setupFailed, shouldExit, err
+	}
+
+	hostname, err := os.Hostname()
+	tc.logger.Execution().Info(errors.Wrap(err, "getting hostname"))
+	if hostname != "" {
+		tc.logger.Execution().Infof("Hostname is '%s'.", hostname)
+	}
+	tc.logger.Task().Infof("Starting task '%s', execution %d.", tc.taskConfig.Task.Id, tc.taskConfig.Task.Execution)
+
+	return tc, setupFailed, shouldExit, nil
```

Comment:
It seems like the value of `setupFailed` is always equivalent to the returned error being non-nil. Can we consolidate the boolean?

Reply:
a.handleTaskResponse overwrites err, so we can't rely on err being non-nil

Reply:
It seems like `setupTask` should return an error though (even if `handleTaskResponse` doesn't produce an error). We can simplify and remove the `setupFailed` boolean by using an error catcher inside `handleSetupError`, catching the error passed into `handleSetupError`, and then catching the error from `handleTaskResponse`, so that returning an error from `setupTask` is equivalent to `setupFailed`.